### ISSUE TITLE
How to define a variable with a restricted domain of strings

### ISSUE THREAD

**Issue by Sushant-Padha** [Wikunia/ConstraintSolver.jl#276]
I am using this constraint solver package for creating a constraint solver for solving timetables with defined constraints on the number of subjects, position of teachers etc. I am rewriting the problem in Julia (was using Python earlier) and calling it in my Python code using [PyCall](https://github.com/JuliaPy/PyCall.jl) (don't wanna rewrite all the class definitions in Julia without OOP).

In the [docs](https://wikunia.github.io/ConstraintSolver.jl/stable/) for the constraint solver, I found this [sample code](https://wikunia.github.io/ConstraintSolver.jl/stable/tutorial/#Graph-coloring) for *defining variables*:
```julia
@variable(m, 1 <= countries[1:5] <= 10, Int)
```

However, I don't understand the syntax `1 <= countries[1:5] <= 10` or its relevance.

In my use case, I want to define a variable with a domain which is a set of strings. Earlier (in Python), I was using this syntax with the [python-constraint](https://github.com/python-constraint/python-constraint) library:
```python3
for variable, domain in zip(variables, domains):
    problem.addVariable(variable, domain)
```
But, I don't know how to translate this into Julia code with this different syntax.

Furthermore, I can not figure out how to use a custom constraint implemented in a function as I could earlier using the python library. The codebase is too complicated for me to understand well, so I would appreciate any help in that regard as well.

Disclaimer: I am pretty new to Julia, so it may be possible that I am just not understanding some simple concept. Also, my understanding of constraint solving in general is very limited.

**Comment by Wikunia** [comment]
Hey, thanks for checking out my solver. As you might have seen this is still in it's infancy as it's quite hard to build a solver from scratch alone 🤣
Especially scheduling constraints are missing for now.

However to your specific questions. The variable definition you mentioned would construct 5 variables `countries[1]` to `countries[5]` where each variable is between 1 and 10 and is an integer. Currently only integer variables can be constructed so you can't directly define a variable for a string. You would need to define some kind of mapping.

You might in general be interested in some more examples: http://hakank.org/julia/constraints/ which are provided by @hakank who experimented a lot with my solver in the last year.

**Comment by Wikunia** [comment]
Regarding variable definition you might want to check the docs for JuMP.jl as well which is the modelling layer below this solver. Let me know if I can help some more ☺️

**Comment by Sushant-Padha** [comment]
Thanks for your input! Really appreciate you helping me.

> Currently only integer variables can be constructed so you can't directly define a variable for a string. You would need to define some kind of mapping.

That is what I thought initially, so I will be implementing that.

But I still don't understand one thing can I create my own constraint as a function?

**Comment by Wikunia** [comment]
There is in theory a way to implement your own constraint. Would love to help you with that as I think it's an important part of the solver but not sure how easy it is to define one from the outside. 
I think I have written down some docs for it or it's on my blog.
https://opensourc.es 
Due to Christmas time I don't have much time to check right now but happy to help next week

**Comment by Sushant-Padha** [comment]
Will look into it, thanks!

**Comment by Sushant-Padha** [comment]
I have simplified all the constraints, variables and values enough so that I can define them using the constructs available in the package. However, I have another problem which I am not sure how to solve.

Basically, I want to define an expression `v` to hold all my variables. Then for a set of variables `v[start:stop]`, I want to define a domain that is made of integers but not continuous, i.e., not a range like `1:10` (which is equivalent to `{1,2,3,4,5,6,7,8,9,10}`) but a specific set like `{1, 3, 4, 6, 7, 8, 10}`.

Can I define that using this syntax? (`vals` is the OrderedSet of values)
```julia
@variable(m, v[start:stop] in vals, Int)
```

Or should I just define all variables with the a range-like domain, and then for each variable remove a set of values? I am thinking that would be a little slower and also counterintuitive.

**Comment by Wikunia** [comment]
Glad that you figured out your way so far. Yes `CS.Integers` is the way to go in this case. i.e 
`@variable(m, y, CS.Integers([1, 4]))`  creates a variable `y` such that only `1` and `4` are possible.

**Comment by Sushant-Padha** [comment]
Thanks for your help!

I have created all the variables succesfully and added teacher constraints (all teachers teaching at the same day and period should be unique) and subject-teacher constraints (a teacher has to actually teach the subject chosen by the solver). But, I do not know how to add subject constraints.

Basically, the subject values chosen should be of a certain number according to a dict. For example, lets say `{1,2,3}` are all the subject values. There exists a mapping like `{1=>2, 2=>5, 3=>3}`, this means that subject `1` should occur exactly 2 times in the timetable, subject `2` should only occur 5 times and subject `3`, 3 times.

In short, I want to define a constraint such that, no matter where a specific value is chosen for variables, it is only done `n` times which can be specified. So, I don't care where value `1` is chosen, but in total it should only be chosen 2 times.

How can I achieve this?

**Comment by Wikunia** [comment]
It depends quite a bit on how you defined your variables. You might want to have a look at the `TableConstraint` if it can't be expressed by a simple sum constraint for example.

**Comment by Sushant-Padha** [comment]
I have defined my variables as a simple vector but using anonymous variables. However, my values are not continuous like `1,2,3` but usually more like `1,3,4`.

I did think of making a list of possible perturbations and adding them with `TableSet`, but quickly found out that even just 10 subjects (in the worst case) will have 10! perturbation which is over 3 million.

I asked this same [question](https://discourse.julialang.org/t/jump-constraint-to-use-a-value-only-n-times/73930) on discourse.julialang.org, and got an answer. Which is that, I should use a binary expansion, with a matrix of binary vars, instead of a vector of int ones. The `j` index of whichever value is set to 1 is my actual int. Using this expression I can define my constraints easily.

**Comment by Wikunia** [comment]
Yeah it's normally easier to work with binary variables. However feel free to write it with the table constraint. 3 million rows should work I think.

**Comment by Sushant-Padha** [comment]
Thanks a lot for all your help, I really appreciate it.

I finally completed my project. Ended up rewriting the full project in Julia. I plan to add it to the registry as `TimetableSolver.jl`.
One example problem that I tried solving in the python version (that took an unknown time to solve more than 12 hours), finished in 0.01 seconds. So, I am pretty satisfied with it.

Have a good day! And, thanks again.

**Comment by Wikunia** [comment]
That makes me extremely happy and excited. Let me know whenever I can be of more help. No matter whether it's questions, ideas, bug reports or performance concerns 😊

Also thanks for GitHub star number 100 😊

**Comment by Sushant-Padha** [comment]
Oh, that's neat, I did not notice that my star was number 100.

Also, I have uploaded my package as [TimetableSolver.jl](https://github.com/Sushant-Padha/TimetableSolver.jl)!  
Really excited for all the cool things I can create with Julia and ConstraintSolver and JuMP.  
This was a great learning experience.
